horovod/mxnet/mpi_ops.cc

```diff
@@ -16,64 +16,53 @@
 #include <chrono>
 #include <memory>
 #include <thread>
+#include <atomic>
 
 #include "../common/operations.h"
 #include "adapter.h"
 #include "cuda_util.h"
-#include "handle_manager.h"
 #include "mpi_ops.h"
 #include "ready_event.h"
 #include "tensor_util.h"
 
 namespace horovod {
 namespace mxnet {
 
-static HandleManager handle_manager;
-
 namespace {
 
+std::atomic_int op_count;
+
 std::string GetOpName(std::string prefix, char* name) {
   if (name != nullptr) {
     return prefix + "." + std::string(name);
   }
-  return std::string();
-}
 
-std::string GetOpNameHandle(std::string prefix, const std::string& name,
-                            int handle) {
-  if (name.length() != 0) {
-    return name;
-  }
-  return prefix + ".noname." + std::to_string(handle);
+  op_count.fetch_add(1);
+  return prefix + ".noname." + std::to_string(op_count);
 }
-
 } // namespace
 
-int DoAllreduce(NDArray* tensor, NDArray* output, const std::string& name,
-                Callback cb) {
+void DoAllreduce(NDArray* tensor, NDArray* output, const std::string& name,
+                 Callback cb) {
   ThrowIfError(common::CheckInitialized());
 
-  auto handle = handle_manager.AllocateHandle(cb);
   auto device = TensorUtil::GetDevice(tensor);
   auto hvd_tensor = std::make_shared<MXTensor<NDArray>>(tensor);
   auto hvd_context = std::make_shared<MXOpContext<NDArray>>(device, output);
   auto hvd_output = std::make_shared<MXTensor<NDArray>>(output);
 
   auto enqueue_result =
       EnqueueTensorAllreduce(hvd_context, hvd_tensor, hvd_output, nullptr,
-                             GetOpNameHandle("allreduce", name, handle), device,
-                             [handle](const Status& status) {
-                               handle_manager.MarkDone(handle, status);
-                               handle_manager.ExecuteCallback(handle);
+                             name, device,
+                             [cb](const Status& status) {
+                               cb();
```

Comment:
I remember in prior code you claimed that it's necessary to have a lock around callback - has it been determined to not be the case?

Reply:
Also, how does errors like shape mismatch get signalled back to user, if we ignore status?

Reply:
The mutex was to protect concurrent access to the unordered_map in handle_manager. shared_ptr was to track the ownership of the callback. We no longer need the lock if we don't return the handle.

Reply:
The lock was needed to synchronize access the unordered_map where callback was stored in the handle manager. We did not need the lock to invoke the callback. We ran tests with this change and did not see issues.

Reply:
If there is error like shape mismatch, exception will be thrown from MXNet engine [here](https://github.com/uber/horovod/pull/542/files/cf0c7c8c3a6633c32146db4b0f4ee4472726e730..24b3f59bb6a821ee2e9a1c926be1a58bfffc5222#diff-79b78539e15db1d75d73029490eed708L237)

Reply:
For some reason, your link doesn't work... Could you try linking against [a line in PR](https://github.com/uber/horovod/pull/542/files) directly? Shape mismatch error originates from Horovod, so `status` would have an error here, but it's getting ignored:
```c++
[cb](const Status& status) {
  cb();
});
```

Reply:
I was referring to [this](https://github.com/uber/horovod/pull/542/files#diff-79b78539e15db1d75d73029490eed708R208) line. However, now I am not sure if I understood your question correctly. Are you suggesting shape mismatch is captured by Horovod and the error status would be written to the `status` variable? Thanks!

Reply:
We did save the Status into enqueue_result and signal errors [here](https://github.com/uber/horovod/pull/542/files#diff-79b78539e15db1d75d73029490eed708R60)

Reply:
That's insufficient - it will only catch a portion of errors.  It will not catch errors discovered by Horovod coordinator (mismatch of shapes between ranks, etc).  Can you try porting tests from PyTorch tests (e.g. `test_horovod_broadcast_type_error`), and seeing the impact?  We should port all the error condition tests.